src/sugar3/graphics/texteditor.py

```diff
@@ -0,0 +1,167 @@
+        self._collab.connect('buddy-joined', self.__buddy_joined_cb)
+        self._collab.setup()
+        self.textview = Gtk.TextView()
+        self.textview.set_editable(True)
+        self.textview.set_cursor_visible(True)
+        self.textview.set_wrap_mode(Gtk.WrapMode.WORD)
+        self.textbuffer = self.textview.get_buffer()
+        self.textbuffer.connect('insert-text', self.__text_buffer_inserted_cb)
+        self.textbuffer.connect('delete-range', self.__text_buffer_deleted_cb)
+        self.textbuffer.set_text("")
+        self.textview.show()
+        self.pack_start(self.textview, expand=True, fill=True, padding=0)
+        self.has_initialized = False
+        self.show()
+
+    '''
+    The message callback is called whenever another user edits
+    something in the text editor and the changes are reflected
+    in the editor or when a new buddy joins and we send them the
+    latest version of the text buffer.
+
+    Args:
+        buddy : another user who sent the message
+        message : updates send over from other users
+    '''
+
+    def __message_cb(self, collab, buddy, message):
+        action = message.get('action')
+        if action == 'init_response' and self.has_initialized == False and message.get('res_id') == self._id:
+            self.has_initialized = True
+            self._callbacks_status = False
+            self.textbuffer.set_text(message.get('current_content'))
+            self._callbacks_status = True
+        if action == 'entry_inserted' and message.get('res_id') == self._id:
+            start_iter=self.textbuffer.get_iter_at_line_offset(message.get('start_iter_line'),
+                    message.get('start_iter_offset'))
+            self._callbacks_status = False
+            self.textbuffer.insert(start_iter, message.get('new_text'))
+            self._callbacks_status = True
+        if action == 'entry_deleted' and message.get('res_id') == self._id:
+            start_iter=self.textbuffer.get_iter_at_line_offset(message.get('start_iter_line'),
+                    message.get('start_iter_offset'))
+            end_iter=self.textbuffer.get_iter_at_line_offset(message.get('end_iter_line'),
+                    message.get('end_iter_offset'))
+            self._callbacks_status = False
+            self.textbuffer.delete(start_iter, end_iter)
+            self._callbacks_status = True
+    
+    '''
+    The buddy joined callback is called whenever another user joins
+    this activity. The leader then send them the contents of the text 
+    buffer so that their text buffer is an identical copy.
+     
+    Args:
+         buddy : another user who has joined the activity
+    '''
+         
+    def __buddy_joined_cb(self, sender, buddy):
+        logging.debug("Buddy joined")
+        if self._collab._leader == False:
+            return
+        if self.has_initialized == False:
+            self.has_initialized = True
+        time.sleep(0.5)
```

Comment:
Hum, that is bad.  Maybe bind to the joined signal, which is called with somebody joins the activity.  You could then send a 'init_request' message and the reply could be send in __message_cb